unitary DFT for symmetric group algebra

jacksonwalters opened the pull request:
<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

- #39200: Uses the decomposition.

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,70 @@ def dft(self, form=None, mult='l2r'):
+                    K = K.extension(sqrt(n).minpoly(),names=gen_name)
+                if not n.is_rational() and sqrt(n).minpoly().is_irreducible():
+                    gen_name = "deg" + str(n.minpoly().degree()) + "index" + str(list(set(required_square_roots)).index(n))
+                    K = K.extension(sqrt(n).minpoly(),names=gen_name)
```
```suggestion
        G = self.group()
        group_size = G.cardinality()
        data = {}
        for partition in Partitions(G.degree()):
            specht_module = self.specht_module(partition)
            rho = specht_module.representation_matrix
            P = sum(rho(g) * rho(g).conjugate().transpose() for g in G) / group_size
            d, L = P.eigenmatrix_left()
            data[partition] = (specht_module, P, d, L)

        def all_roots_field():
            required_square_roots = []
            for partition in Partitions(G.degree()):
                specht_module, P, d, L = data[partition]
                required_square_roots += [specht_module.dimension(), group_size] + d.diagonal()
            required_square_roots = flatten([[QQ(q).numerator(), QQ(q).denominator()] if q in QQ else q for q in required_square_roots])
            K = self.base_ring()
            for n in set(required_square_roots):
                R = PolynomialRing(K, 'x')
                x = R.gen()
                if n.is_rational() and (x**2 - n).is_irreducible():
                    gen_name = "sqrt"+str(n).replace("/","over")
                    K = K.extension(sqrt(n).minpoly(),names=gen_name)
                if not n.is_rational() and sqrt(n).minpoly().is_irreducible():
                    gen_name = "deg" + str(n.minpoly().degree()) + "index" + str(list(set(required_square_roots)).index(n))
                    K = K.extension(sqrt(n).minpoly(),names=gen_name)
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,70 @@ def dft(self, form=None, mult='l2r'):
+            group_size = self.group().cardinality()
+            P = (1/group_size)*sum(rho(g)*rho(g).conjugate().transpose() for g in self.group())
+            d, L = P.eigenmatrix_left()
+            return L.inverse() * diagonal_matrix([sqrt(K(a)) for a in d.diagonal()]) * L
```
```suggestion
        def unitary_change_of_basis(partition, K):
            specht_module, P, d, L = data[partition]
            return L.inverse() * diagonal_matrix([sqrt(K(a)) for a in d.diagonal()]) * L
        K = all_roots_field()
        unitary_data = {partition: unitary_change_of_basis(partition, K) for partition in Partitions(G.degree())}
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,70 @@ def dft(self, form=None, mult='l2r'):
+            Q = unitary_change_of_basis(partition,K)
+            unitary_factor = specht_module.dimension()/self.group().cardinality()
+            sqrt_unitary_factor = sqrt(K(unitary_factor))
+            return sqrt_unitary_factor*sum(f(g)*Q.inverse()*rho(g)*Q for g in self.group())
```
```suggestion
        def hat(f, partition):
            specht_module, P, d, L = data[partition]
            rho = self.specht_module(partition).representation_matrix
            Q = unitary_data[partition]
            unitary_factor = specht_module.dimension() / group_size
            sqrt_unitary_factor = sqrt(K(unitary_factor))
            return sqrt_unitary_factor * sum(f(g) * Q.inverse() * rho(g) * Q for g in G if f(g))
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,70 @@ def dft(self, form=None, mult='l2r'):
+
+        K = all_roots_field()
+        delta = lambda s: lambda t: 1 if t == s else 0
+        return matrix(K,[flatten([hat(delta(g),partition,K).list() for partition in Partitions(self.group().degree())]) for g in self.group()]).transpose()
```
```suggestion
        return matrix(K, [flatten([hat(delta(g), partition).list() for partition in Partitions(G.degree())]) for g in G]).transpose()
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,70 @@ def dft(self, form=None, mult='l2r'):
+            sqrt_unitary_factor = sqrt(K(unitary_factor))
+            return sqrt_unitary_factor*sum(f(g)*Q.inverse()*rho(g)*Q for g in self.group())
+
+        K = all_roots_field()
```
```suggestion
```
Moved earlier.

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,70 @@ def dft(self, form=None, mult='l2r'):
+            [               0       -1/3*sqrt3        1/6*sqrt3              1/2             -1/2        1/6*sqrt3]
+            [               0       -1/3*sqrt3        1/6*sqrt3             -1/2              1/2        1/6*sqrt3]
+            [      -1/3*sqrt3                0             -1/2        1/6*sqrt3        1/6*sqrt3              1/2]
+            [-1/6*sqrt2*sqrt3  1/6*sqrt2*sqrt3  1/6*sqrt2*sqrt3 -1/6*sqrt2*sqrt3 -1/6*sqrt2*sqrt3  1/6*sqrt2*sqrt3]
```
Can you add a test showing the unitary property?

jacksonwalters replied:
Added U_dft*U_dft.conjugate().transpose() == identity_matrix(SGA.group().cardinality())
True

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                return sqrt_unitary_factor * sum(f(g) * Q.inverse() * rho(g) * Q for g in G if f(g))
+
+            delta = lambda s: lambda t: 1 if t == s else 0
+            return matrix(K, [flatten([hat(delta(g), partition).list() for partition in Partitions(G.degree())]) for g in G]).transpose()
```
```suggestion
            def hat(g, partition):
                specht_module, P, d, L = data[partition]
                rho = self.specht_module(partition).representation_matrix
                Q = unitary_data[partition]
                unitary_factor = specht_module.dimension() / G.cardinality()
                sqrt_unitary_factor = sqrt(K(unitary_factor))
                return sqrt_unitary_factor * Q.inverse() * rho(g) * Q

            PGdeg = Partitions(G.degree())
            return matrix(K, [flatten([hat(g, partition).list() for partition in PGdeg]) for g in G]).transpose()
```
I think this is equivalent since `delta` is just getting the summand `g` .

Although you probably can write this to get rid of that transpose and the `flatten`... The partitions parent `PGdeg` can also be created earlier (and reused here).

jacksonwalters replied:
Yes, this was written to be general and obtain the Fourier coefficient for any function f. Since we only need the summand `g`, this is simpler.

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                        K = K.extension(sqrt(n).minpoly(),names=gen_name)
+                    if not n.is_rational() and sqrt(n).minpoly().is_irreducible():
+                        gen_name = "deg" + str(n.minpoly().degree()) + "index" + str(list(set(required_square_roots)).index(n))
+                        K = K.extension(sqrt(n).minpoly(),names=gen_name)
```
```suggestion
                        K = K.extension(sqrt(n).minpoly(), names=gen_name)
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                for g in G:
+                    total_system = total_system.stack(augmented_matrix(g))
+                null_space = total_system.right_kernel()
+                U_mats = [matrix(F,d_rho,d_rho,b) for b in null_space.basis()]
```
```suggestion
                U_mats = [matrix(F, d_rho, d_rho, b) for b in null_space.basis()]
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                            row = [linear_expression.coefficient(u) for u in U_vars]
+                            augmented_system.append(row)
+                    return matrix(F, augmented_system)
+                total_system = matrix(F,0,d_rho**2)
```
```suggestion
                total_system = matrix(F, 0, d_rho**2)
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                d_rho = specht_module.dimension()
+                R = PolynomialRing(F,'u',d_rho**2)
+                U_vars = R.gens()
+                U = matrix(R,d_rho,d_rho,U_vars)
```
```suggestion
                R = PolynomialRing(F, 'u', d_rho**2)
                U_vars = R.gens()
                U = matrix(R, d_rho, d_rho, U_vars)
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                z = F.multiplicative_generator()
+                k = u.log(z)
+                if k % (q+1) != 0:
+                    raise ValueError("Unable to factor: u is not in base field GF(q)")
```
```suggestion
                    raise ValueError(f"unable to factor since {u} is not in base field GF({q})")
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                k = u.log(z)
+                if k % (q+1) != 0:
+                    raise ValueError("Unable to factor: u is not in base field GF(q)")
+                return z**((k//(q+1)) % (q-1))
```
```suggestion
                return z ** ((k//(q+1)) % (q-1))
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+                return sqrt_unitary_factor*sum(f(g)*A*rho(g)*A.inverse() for g in G)
+
+            delta = lambda s: lambda t: 1 if t == s else 0
+            fourier_transform = [flatten([hat(delta(g),partition).list() for partition in Partitions(G.degree())]) for g in G]
```
```suggestion
            def hat(g, partition):
                specht_module = self.specht_module(partition)
                rho = specht_module.representation_matrix
                U = invariant_symmetric_bilinear_matrix(partition)[0]
                A = base_change_hermitian(U)
                sqrt_unitary_factor = sqrt(F(specht_module.dimension() / G.cardinality()))
                return sqrt_unitary_factor * f(g) * A * rho(g) * A.inverse()

            PGdeg = Partitions(G.degree())
            fourier_transform = [flatten([hat(g, partition).list() for partition in PGdeg]) for g in G]
```
Same comment as the char 0 case. Although it would likely be better to cache those square root computations somehow.

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+            dft_matrix = matrix(F,fourier_transform).transpose()
+            sign_diag = (dft_matrix*dft_matrix.H).diagonal()
+            factor_diag = diagonal_matrix([conj_square_root(d) for d in sign_diag])
+            return factor_diag.inverse()*dft_matrix
```
```suggestion
            dft_matrix = matrix(F, fourier_transform).transpose()
            sign_diag = (dft_matrix * dft_matrix.H).diagonal()
            factor_diag_inv = diagonal_matrix([~conj_square_root(d) for d in sign_diag])
            return factor_diag_inv * dft_matrix
```
Inverting a diagonal matrix is easy, and this should be faster.

jacksonwalters replied:
Agree, inverting each diagonal element with ~ is better.

jacksonwalters committed:
remove debugging tests

check if mult. gen. and seminormal DFT match

check to see if the multiplicative generator of GF(25) being computed by Sage matches the one being computed locally. also check the DFT. these are the only two things that may be different.

remove double colon in examples

I'm using some descriptive strings for each example since they're all a bit different. those sentences have double colons after them, and code immediately follows.

Update src/sage/combinat/symmetric_group_algebra.py

remove whitespace between `p` and \mid

improve doctest examples with better exposition

say what the example is before performing it. check that the desired property holds.

fix failing doctest examples

the computed unitary DFT in Sage is different than when I compute it locally using the same code. not sure why.

Update src/sage/matrix/matrix2.pyx

fetch the cached rank. if it's full or not yet computed, continue, otherwise throw a ValueError that the matrix is not full rank.

better naming convention for SGAs

separate the field and the group with an underscore.  for fields of size q**2, use the actual squared value

remove extraneous ")" chars

not sure why these are there.

should be GF(9), a field of square order

this is meant to test the case when the characteristic divides the order. in this case, p=3, q=9, and n=3. before we were working over GF(3) rather than GF(9), so it triggered the other checks re: whether it is a field of square order.

remove unnecessary blank lines, duplicate check

remove trailing whitespace

update `form` input, add more tests

include a description in DFT of the different possible values of `form`. the possible values are `None`, `seminormal`, `modular`, `unitary`. each uses a different basis.

also includes new tests for the four different ways that the input field might fail to be a finite field of square order such that the characteristic does not divide the order of the group.

Update src/sage/combinat/symmetric_group_algebra.py

use lowercase for the ValueError message

trigger CI tests

use cholesky(extended=True)

we have renamed hermitian_decomposition to cholesky(extended=True)

Update src/sage/combinat/symmetric_group_representations.py

move .H to cached part

Update src/sage/combinat/symmetric_group_representations.py

move conjugate-transpose to cached part

update doctests

we're using .H in the unitary_change_of_basis, so the doctests need to reflect that

partition, not self._partition

forgot underscore, don't define _n

define self._n

use dimension instead

when using the SGA version of specht, we don't have access to the yang_baxter_graph

use self._specht with SGA specht module

use representation from SGA

using self._specht causes tests to fail

remove print null_space.basis()

appears to be working now with the change to using the symmetric group algebra to get the representation

remove print statements

use SGA vs. _spect for now

include another doctest, print

try printing U

use A.H instead of A

since U = A*A.H, we need to adjust the unitary representation to be A.H\rho(g)A.H.inverse().

also need to update the doctests

remove unnecessary import, q, whitespace

trigger CI checks

use hermitian_decomposition in unitary rep'n

we now have a standalone method to compute the Hermitian decomposition of a matrix, so we should use it when computing the change-of-basis matrix for a unitary representation

implement unitary DFT of symmetric group over finite fields

unitary DFT of symmetric group for number fields and finite fields

- perform the unitary DFT of the symmetric group over finite fields and number fields
- compute unitary representations of the symmetric group over finite fields
- use real orthogonal representations as unitary representations for symmetric group

trigger CI checks

Update src/sage/matrix/matrix2.pyx

add a SEEALSO since this is an extended form of the Cholesky decomposition

Update src/sage/matrix/matrix2.pyx

remove this break conditional in favor of conditional for while loop

Update src/sage/matrix/matrix2.pyx

replace always True with explicit conditional

move hermitian_decomposition into cholesky

this method for decomposition a Hermitian matrix U = AA* is similar in spirit to the Cholesky decomposition, but extends it to work over finite fields of square order.

Update src/sage/matrix/matrix2.pyx

yes, just forgot to update this

initialize row to -1

Update src/sage/matrix/matrix2.pyx

add word "package" to GAP ``forms``

Update src/sage/matrix/matrix2.pyx

change \ast to * for consistency

Trigger CI tests

Trigger CI tests

Update src/sage/matrix/matrix2.pyx

use Matrix

handle 1x1 case

Update src/sage/matrix/matrix2.pyx

cache the rank of the matrix since are computing it anyways

add colon after Traceback

forgot the colon

fail cases should be tests, not examples

Update src/sage/matrix/matrix2.pyx

add hermitian_decomposition method

given a Hermitian matrix U, returns a matrix A such that U = AA*

use translated GAP source code for BaseChangeCanonical

raise ValueError for non full rank matrices

this method does not work for singular matrices. since we are already computing the rank in the row reduction, we can just check at the end whether the rank is full, and if not exit with a ValueError. this may be overcautious - it's not clear that this won't work for some singular matrices. there are cases where it certainly doesn't work, and we include one in the doctest.

add example for singular case

this example is currently failing and needs to be caught

remove unnecessary import, singularity check

if the matrix `U` is singular, the process will still result in a matrix but it will fail to have the expected property, namely `B*B.H == U`.

Update src/sage/matrix/matrix2.pyx

avoid import, directly call `sqrt` method, don't extend to keep result in `ZZ`

fix example

this was using the output from the `forms` package, when it should be from the GAP source translation.

change examples

there is a failing example due to the fact that the matrix is singular with q=3, and U = matrix(F,[[1,4,7],[4,1,4],[7,4,1]]).

move _cholesky_extended_ff to hidden method

move _cholesky_extended_ff to a separate hidden method, and just call it from inside cholesky if extended=true. add doctests and docs for the method separately. sparse for now, should be expanded upon.

Update src/sage/matrix/matrix2.pyx

add output for GF(3**2) case

this case fails to have a lower triangular decomposition which has been checked by brute force.

Update src/sage/matrix/matrix2.pyx

move two examples over finite fields from tests

the two examples computing the extended Cholesky decomposition over square order finite fields are in TESTS, and should be moved to EXAMPLES.

Update src/sage/matrix/matrix2.pyx

this is better and clarifies that the result of the Cholesky decomposition is lower triangular (and not possibly upper triangular).

Update src/sage/matrix/matrix2.pyx

add single whitespace before `extended` parameter

Update src/sage/matrix/matrix2.pyx

looks good. makes sense to point out the Cholesky decomposition might not exist, given that was our initial difficulty.

3 trailing whitespace

move hermitian_decomposition into cholesky

this method for decomposition a Hermitian matrix U = AA* is similar in spirit to the Cholesky decomposition, but extends it to work over finite fields of square order.

add two examples

add two examples of a Hermitian decomposition, one of which is not upper/lower triangular (so would be impossible with Cholesky decomposition)

Update src/sage/matrix/matrix2.pyx

yes, just forgot to update this

initialize row to -1

Update src/sage/matrix/matrix2.pyx

add word "package" to GAP ``forms``

Update src/sage/matrix/matrix2.pyx

change \ast to * for consistency

Trigger CI tests

use self.__class__ within matrix.pyx file

Trigger CI tests

remove blank line

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

use Matrix

handle 1x1 case

format failing test case properly

format failing test case properly by removing variables that hold results of computation. just call the function that raises the exception

Update src/sage/combinat/symmetric_group_representations.py

formatting changes (use ^T rather than .T, use `\rho` instead of `rho`).

Update src/sage/combinat/symmetric_group_representations.py

formatting changes

Update src/sage/combinat/symmetric_group_representations.py

use `.. MATH::` formatting instead of inline code.

Update src/sage/combinat/symmetric_group_representations.py

some formatting changes. add a return, and remove "this is".

move docstring from __init__ to class level

the description of the mathematics should move from the __init__ function which should only contain "Initialize `self`".

whitespace, math formatting issues

add unitary DFT examples

add some examples with `q=5` and `q=7` and `n=3` for the unitary DFT.

add unitary DFT example, case for p|n

include an example of the unitary DFT under the `dft` method (rather than just _dft_unitary). include a test for a case GF(4)S_3, when 2|3!, so that the unitary DFT throws a NotImplementedError.

Update src/sage/combinat/symmetric_group_algebra.py

separate each `form` type into three separate bullets, and don't include None

remove extra line from debugging

add space

update doctests with correct outputs

it should just say the partition for UnitaryRepresentation. change 3 to 4 since the partition is [3,1]

Update src/sage/combinat/symmetric_group_representations.py

change 3 to 4 for correct partition size

Update src/sage/combinat/symmetric_group_representations.py

remove unmatched right parentheses

remove whitespace

add unitary for SymmetricGroupRepresentation(s)

add an input of type `unitary` for the SymmetricGroupRepresentation(s) constructors

Update src/sage/combinat/symmetric_group_algebra.py

optimization for multiplying the diagonal in the number field case

Update src/sage/combinat/symmetric_group_algebra.py

Update src/sage/combinat/symmetric_group_algebra.py

perform the matrix multiplcation and diagonal matrix multiplcation in the one loop

Update src/sage/combinat/symmetric_group_algebra.py

replace `\mid` with `|`

Update src/sage/combinat/symmetric_group_algebra.py

add tick for seminormal

Co-Authored-By: Travis Scrimshaw <clfrngrown@aol.com>
Co-Authored-By: Dima Pasechnik <dimpase@gmail.com>

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -999,6 +1001,170 @@ def _repr_(self):
+        assert self._ring.characteristic() > 0
+        rho = self._specht.representation_matrix
+        A = self._unitary_change_basis_matrix
+        return A.H * rho(permutation) * A.H.inverse()
```
```suggestion
        A = self._unitary_change_basis_matrix
        return A * rho(permutation) * A.inverse()
```

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -999,6 +1001,170 @@ def _repr_(self):
+        total_system = sum((augmented_matrix(g) for g in G), [])
+        null_space = matrix(F, total_system).right_kernel()
+        U = matrix(F, d_rho, d_rho, null_space.basis()[0])
+        return U.hermitian_decomposition()
```
```suggestion
        return U.hermitian_decomposition().H
```

tscrim reviewed (changes requested):
Moving the conjugate-transpose part into the cached portion.

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2066,8 +2066,74 @@ def dft(self, form=None, mult='l2r'):
+            z = F.multiplicative_generator()
+            k = u.log(z)
+            if k % (q+1) != 0:
+                raise ValueError(f"Unable to factor as {u} is not in base field GF({q})")
```
```suggestion
                raise ValueError(f"unable to factor as {u} is not in base field GF({q})")
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2066,8 +2066,74 @@ def dft(self, form=None, mult='l2r'):
+
+        # positive characteristic case
+        assert F.characteristic() > 0
+        if not (F.is_field() and F.is_finite() and F.order().is_square()):
```
Please add some tests checking these conditions on the field.

jacksonwalters replied:
This is addressed in commit [b33cefade43e651454b2c0fbb01adee86cff94a7](https://github.com/jacksonwalters/sage/commit/b33cefade43e651454b2c0fbb01adee86cff94a7). I added four tests. 

Note the assert will always pass, since if a ring is not characteristic zero, it must be positive characteristic.

tscrim reviewed (changes requested):
Please also add to the documentation of `dft` about all of the `form` inputs (I should have caught this on the modular DFT PR...).

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2035,6 +2035,15 @@ def dft(self, form=None, mult='l2r'):
+            of the group algebra into blocks with central orthogonal idempotents.
+            If set to `unitary`, we use the unitary DFT which computes the extended
+            Cholesky decomposition of an S_n-invariant symmetric bilinear form as
+            a change-of-basis matrix over a finite field of square order.
```
```suggestion
        - ``form`` -- string (default: ``"modular"`` if `p |n!` else ``"seminormal"``);
          one of the following:

          * `"seminormal"` -- use the seminormal basis
          * ``"modular"`` -- use the modular DFT, which uses the Peirce decomposition
            of the group algebra into blocks with central orthogonal idempotents
          * ``"unitary"`` -- use the unitary DFT. which computes the extended
            Cholesky decomposition of an `S_n`-invariant symmetric bilinear form as
            a change-of-basis matrix (for positive characteristics, it must be
            a finite field of square order)
```

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -999,6 +1001,170 @@ def _repr_(self):
+# #### Unitary Representation ###############################################
+
+
+class UnitaryRepresentation(SymmetricGroupRepresentation_generic_class):
```
Please add documentation here describing the mathematics.

jacksonwalters replied:
I added some documentation for `__init__`, `_unitary_change_basis_matrix`, and `_representation_matrix_uncached`. I try to capture the three key computations: 1) computing a G-invariant symmetric bilinear form `U` 2) computing its extended Cholesky decomposition `A` 3) check basis-change by `A` is unitary.

tscrim replied:
Please move the `__init__` docstring to the class-level. The `__init__` docstring should just be "Initialize \`\`self\`\`" with the examples. The class-level doesn't need examples, but it should have the mathematics since it is public readable.

jacksonwalters replied:
Makes sense. I've moved the docstring from `__init__` up to the class level.

tscrim reviewed (changes requested):
Thank you. Three things left. Two are comments below and the other is add the examples `QS3.dft(form="unitary")` and `GF4S3.dft(form="unitary")` (note I changed the field to `GF(4)` here).

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -999,6 +1001,186 @@ def _repr_(self):
+class UnitaryRepresentation(SymmetricGroupRepresentation_generic_class):
+    r"""
+    This is a unitary representation of the symmetric group.
+    In characteristic zero, this is the same as the orthogonal representation since
```
```suggestion
    r"""
    A unitary representation of the symmetric group.

    In characteristic zero, this is the same as the orthogonal representation since
```
Thanks. Here is one formatting change.

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -999,6 +1001,186 @@ def _repr_(self):
+        Return the matrix representing the ``permutation`` in this
+        irreducible representation. We use the unitary change of basis
+        matrix to conjugate the representation matrix. These are unitary
+        since `(A * rho(g) * A.inverse()).H * (A * rho(g) * A.inverse()) = A.inverse().H * rho(g).H * A.H * A * rho(g) * A.inverse() = A.inverse().H * rho(g).T * U * rho(g) * A.inverse() = A.inverse().H * U * A.inverse() = 1`.
```
```suggestion
        Return the matrix representing the ``permutation`` in this
        irreducible representation.

        We use the unitary change of basis matrix to conjugate
        the representation matrix. These are unitary since

        .. MATH::

            (A rho(g) A^{-1})^* \cdot (A rho(g) A^{-1})
              = (A^{-1})^* rho(g)^* A^* A rho(g) A^{-1}
              = (A^{-1})^*H rho(g)^T U rho(g) A^{-1}
              = (A^{-1})^* U A^{-1} = 1.
```

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -999,6 +1001,186 @@ def _repr_(self):
+    def _unitary_change_basis_matrix(self):
+        """
+        Compute the change of basis matrix. We first compute a G-invariant symmetric bilinear form. This
+        yields a solution to the equation `rho(g).T * U * rho(g) = U` for all `g` in `G`. We can
```
```suggestion
        Compute the change of basis matrix.

        We first compute a `G`-invariant symmetric bilinear form. This yields a solution `U`
        to the equation `\rho(g)^T U \rho(g) = U` for all `g` in `G`. We can
```

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -999,6 +1001,186 @@ def _repr_(self):
+    representations when the field is finite, has square order, and the characteristic does not divide
+    the order of the group. In this case, we construct the representation by first
+    constructing the orthogonal representation and then taking the extended Cholesky
+    decomposition of the unique solution to the equation `rho(g).T * U * rho(g) = U` for all `g` in `G`.
```
```suggestion
    constructing the orthogonal representation `\rho` and then taking the extended
    Cholesky decomposition of the unique solution `U` to the equation
    `\rho(g)^T U \rho(g) = U` for all `g` in `G`.
```

tscrim reviewed (approved):
Thank you. Some additional formatting things for readability.

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2035,28 +2035,57 @@ def dft(self, form=None, mult='l2r'):
           (:meth:`antipode`) of the seminormal basis instead of
           the seminormal basis.
 
+        - ``form`` -- string (default: ``"modular"`` if `p |n!` else ``"seminormal"``);
```
```suggestion
        - ``form`` -- string (default: ``"modular"`` if `p|n!` else ``"seminormal"``);
```

tscrim replied:
Or the other way
```suggestion
        - ``form`` -- string (default: ``"modular"`` if `p | n!` else ``"seminormal"``);
```

tscrim reviewed (changes requested):
Okay, one last mystery to solve. Why do you get a different ordering of the columns than the bots. I don't see where there could be something that produces different orders. Tests are passing locally for you?

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2035,22 +2035,60 @@ def dft(self, form=None, mult='l2r'):
 
-            sage: QS3 = SymmetricGroupAlgebra(QQ, 3)
-            sage: QS3.dft()
+          * `"seminormal"` -- use the seminormal basis
```
```suggestion
          * ``"seminormal"`` -- use the seminormal basis
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2035,22 +2035,60 @@ def dft(self, form=None, mult='l2r'):
+            sage: U*U.H == 1
+            True
+
         Over fields of characteristic `p > 0` such that `p \mid n!`, we use the
```
```suggestion
        Over fields of characteristic `p > 0` such that `p|n!`, we use the
```
Just to make it uniform. Or you can change the one you added.

tscrim reviewed (changes requested):
Sorry, I know my comments are coming in small batches over time rather than just being done with it all. This should be the last round though.

Two minor formatting things, plus you should add the `"unitary"` input to the `SymmetricGroupRepresentations` docstring with an example.

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2066,8 +2104,103 @@ def dft(self, form=None, mult='l2r'):
+        dft_matrix = self.dft()
+        sign_diag = (dft_matrix * dft_matrix.H).diagonal()
+        conj_sqrt_diag_inv = diagonal_matrix([~conj_square_root(d) for d in sign_diag])
+        return conj_sqrt_diag_inv * dft_matrix
```
```suggestion
        n = dft_matrix.nrows()
        for i in range(n):
            d = sum(dft_matrix[i, j] * dft_matrix[i, j].conjugate() for j in range(n))
            dft_matrix[i] *= ~conj_square_root(d)
        return dft_matrix
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2066,8 +2104,103 @@ def dft(self, form=None, mult='l2r'):
+            from sage.misc.functional import sqrt
+            from sage.rings.number_field.number_field import NumberField
+            dft_matrix = self.dft()
+            diag = (dft_matrix * dft_matrix.H).diagonal()
```
```suggestion
            n = dft_matrix.nrows()
            diag = [sum(dft_matrix[i, j] * dft_matrix[i, j].conjugate() for j in range(n))
                    for i in range(n)]
```

tscrim commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -2066,8 +2104,103 @@ def dft(self, form=None, mult='l2r'):
+            x = PolynomialRing(QQ, 'x').gen()
+            K = NumberField([x**2 - d for d in primes_needed], names=names)
+            sqrt_diag_inv = diagonal_matrix([~sqrt(K(d)) for d in diag])
+            return sqrt_diag_inv * dft_matrix
```
```suggestion
            dft_matrix = dft_matrix.change_ring(K)
            for i, d in enumerate(diag):
                dft_matrix[i] *= ~sqrt(K(d))
            return dft_matrix
```

tscrim reviewed (changes requested):
Sorry, now I am doing some obvious optimizations. This will be it from me. Anything else I can do on a followup PR.

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -238,6 +257,23 @@ def SymmetricGroupRepresentations(n, implementation='specht', ring=None,
         [ 0  0  0 -1  1]
         [ 0  1 -1 -1  1]
         [ 0  1  0 -1  1]
+    
```
```suggestion

```
Trailing whitespace.

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -131,6 +132,23 @@ def SymmetricGroupRepresentation(partition, implementation='specht',
         [ 0  1  0 -1  1]
         sage: spc.verify_representation()
         True
+    
```
```suggestion

```
Trailing whitespace.

tscrim reviewed (approved):
Thank you. Just two last details for the linter.

The bottleneck is now in computing the seminormal form, which seems like a harder problem to optimize (with a fair bit of the time spent in multiplying elements in the SGA).

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -239,6 +258,23 @@ def SymmetricGroupRepresentations(n, implementation='specht', ring=None,
+
+        sage: unitary = SymmetricGroupRepresentations(3, "unitary"); unitary
+        Unitary representations of the symmetric group of order 3! over Symbolic Ring
+        sage: unitary_GF49 = SymmetricGroupRepresentations(3, "unitary", ring=GF(7**2)); unitary_GF49
```
```suggestion
        sage: unitary_GF49 = SymmetricGroupRepresentations(4, "unitary", ring=GF(7**2)); unitary_GF49
```

tscrim commented on src/sage/combinat/symmetric_group_representations.py:
```diff
@@ -239,6 +258,23 @@ def SymmetricGroupRepresentations(n, implementation='specht', ring=None,
+        [6 0 0]
+        [0 1 0]
+        [0 0 1]
+        sage: unitary_GF49([3,1])([3,2,1,4]))
```
```suggestion
        sage: unitary_GF49([3,1])([3,2,1,4])
```

tscrim reviewed (commented):
This should address the doctests.
